core/src/main/scala/cats/data/NonEmptyVector.scala

```diff
@@ -3,69 +3,99 @@ package data
+  /** Gets the element at the index, if it exists */
+  def get(i: Int): Option[A] =
+    Try(getUnsafe(i)).toOption
 
-  def head: A = vector.head
+  /** Gets the element at the index, or throws an exception if none exists */
+  def getUnsafe(i: Int): A = toVector(i)
 
-  def tail: Vector[A] = vector.tail
+  /** Updates the element at the index, if it exists */
+  def updated(i: Int, a: A): Option[NonEmptyVector[A]] =
+    Try(updatedUnsafe(i, a)).toOption
+
+  /** Updates the element at the index, or throws an exeption if none exists */
+  def updatedUnsafe(i: Int, a: A):
+      NonEmptyVector[A] = NonEmptyVector(toVector.updated(i, a))
+
+  def head: A = toVector.head
+
+  def tail: Vector[A] = toVector.tail
 
   /**
    * remove elements not matching the predicate
    */
-  def filter(f: A => Boolean): Vector[A] = vector.filter(f)
+  def filter(f: A => Boolean): Vector[A] = toVector.filter(f)
 
   /**
    * Append another NonEmptyVector to this
    */
-  def combine(other: NonEmptyVector[A]): NonEmptyVector[A] = NonEmptyVector(vector ++ other.vector)
+  def concat(other: NonEmptyVector[A]): NonEmptyVector[A] = NonEmptyVector(toVector ++ other.toVector)
+
+  /**
+   * Append another Vector to this
+   */
+  def concat(other: Vector[A]): NonEmptyVector[A] = NonEmptyVector(toVector ++ other)
 
   /**
    * find the first element matching the predicate, if one exists
    */
-  def find(f: A => Boolean): Option[A] = vector.find(f)
+  def find(f: A => Boolean): Option[A] = toVector.find(f)
 
   /**
    * Check whether at least one element satisfies the predicate.
    */
-  def exists(f: A => Boolean): Boolean = vector.exists(f)
+  def exists(f: A => Boolean): Boolean = toVector.exists(f)
 
   /**
    * Check whether all elements satisfy the predicate.
    */
-  def forall(f: A => Boolean): Boolean = vector.forall(f)
+  def forall(f: A => Boolean): Boolean = toVector.forall(f)
 
   /**
    * Left-associative fold using f.
    */
   def foldLeft[B](b: B)(f: (B, A) => B): B =
-    vector.foldLeft(b)(f)
+    toVector.foldLeft(b)(f)
 
   /**
    * Right-associative fold using f.
    */
   def foldRight[B](lb: Eval[B])(f: (A, Eval[B]) => Eval[B]): Eval[B] =
-    Eval.defer(Foldable[Vector].foldRight(vector, lb)(f))
+    Foldable[Vector].foldRight(toVector, lb)(f)
 
   /**
     * Applies f to all the elements
     */
   def map[B](f: A => B): NonEmptyVector[B] =
-    NonEmptyVector(vector.map(f))
+    NonEmptyVector(toVector.map(f))
 
   /**
     *  Applies f to all elements and combines the result
     */
   def flatMap[B](f: A => NonEmptyVector[B]): NonEmptyVector[B] =
-    NonEmptyVector(vector.flatMap(a => f(a).vector))
+    NonEmptyVector(toVector.flatMap(a => f(a).toVector))
+
+  /**
+    * Left-associative reduce using f.
+    */
+  def reduceLeft(f: (A, A) => A): A =
+    tail.foldLeft(head)(f)
+
+  /**
+    * Left-associative reduce using the Semigroup of A
+    */
+  def reduce(implicit S: Semigroup[A]): A =
+    reduceLeft(S.combine)
```

Comment:
semigroup has `combineOption` which can possibly be faster than using `combine`.